src/operations/execute_operation.ts

```diff
@@ -94,63 +96,51 @@ export function executeOperation<
         owner = Symbol();
         session = topology.startSession({ owner, explicit: false });
       } else if (session.hasEnded) {
-        return cb(new MongoExpiredSessionError('Use of expired sessions is not permitted'));
+        return callback(new MongoExpiredSessionError('Use of expired sessions is not permitted'));
       } else if (session.snapshotEnabled && !topology.capabilities.supportsSnapshotReads) {
-        return cb(new MongoCompatibilityError('Snapshot reads require MongoDB 5.0 or later'));
+        return callback(new MongoCompatibilityError('Snapshot reads require MongoDB 5.0 or later'));
       }
     } else if (session) {
       // If the user passed an explicit session and we are still, after server selection,
       // trying to run against a topology that doesn't support sessions we error out.
-      return cb(new MongoCompatibilityError('Current topology does not support sessions'));
+      return callback(new MongoCompatibilityError('Current topology does not support sessions'));
     }
 
     try {
-      executeWithServerSelection(topology, session, operation, (err, result) => {
-        if (session && session.owner && session.owner === owner) {
-          return session.endSession(err2 => cb(err2 || err, result));
+      executeWithServerSelection<TResult>(topology, session, operation, (error, result) => {
+        if (session?.owner != null && session.owner === owner) {
+          return session.endSession(endSessionError => callback(endSessionError ?? error, result));
         }
 
-        cb(err, result);
+        callback(error, result);
       });
-    } catch (e) {
-      if (session && session.owner && session.owner === owner) {
+    } catch (error) {
+      if (session?.owner != null && session.owner === owner) {
         session.endSession();
       }
-
-      throw e;
+      throw error;
     }
   });
 }
 
-function supportsRetryableReads(server: Server) {
-  return maxWireVersion(server) >= 6;
-}
-
-function executeWithServerSelection(
+function executeWithServerSelection<TResult>(
   topology: Topology,
   session: ClientSession,
   operation: AbstractOperation,
-  callback: Callback
+  callback: Callback<TResult>
 ) {
-  const readPreference = operation.readPreference || ReadPreference.primary;
-  const inTransaction = session && session.inTransaction();
+  const readPreference = operation.readPreference ?? ReadPreference.primary;
+  const inTransaction = !!session?.inTransaction();
 
   if (inTransaction && !readPreference.equals(ReadPreference.primary)) {
-    callback(
+    return callback(
       new MongoTransactionError(
         `Read preference in a transaction must be primary, not: ${readPreference.mode}`
       )
     );
-
-    return;
   }
 
-  if (
-    session &&
-    session.isPinned &&
-    session.transaction.isCommitted &&
-    !operation.bypassPinningCheck
-  ) {
+  if (session?.isPinned && session?.transaction.isCommitted && !operation.bypassPinningCheck) {
```

Comment:
How come the optional chaining is necessary here?  Session is of ClientSession and shouldn't be undefined, so we shouldn't need to check for its existence.

Reply:
Apparently this was the source of the CI difficulties I've been having, despite TS claiming the session is always present removing the checks for existence fail a ton of the CSFLE tests but they only fail on evergreen. 🤔